Branch-after-load scheme for OCaml memory model on Arm64 #2

dustanddreams opened the pull request:
This PR implements the alternate code generation scheme for arm64 described in #13393, but as a compiler option, allowing users to try both schemes and pick the one which works best for their workload.

The option is named `-falternate-code-generation`. 

Missing at the moment of PR creation (hence the draft status):
- consensus on the option name.
- how to store the state of this option so that mixing object files built with different schemes is not allowed.
- `Changes` entry.
- documentation for that option.

Poke @kayceesrk

ghost commented:
Mentioning #13262 as well.

gasche commented:
> consensus on the option name.

My proposal: `-barrier-placement=[fence-before-store | branch-after-load]`.

kayceesrk commented on asmcomp/arm64/emit.mlp:
```diff
@@ -464,23 +464,46 @@ module BR = Branch_relaxation.Make (struct
-      based + barrier + single
+    | Lop (Iload  { memory_chunk; addressing_mode; is_atomic; mutability }) ->
+      if !Clflags.alternate_code_generation then begin
+        let based = match addressing_mode with Iindexed _ -> 0 | Ibased _ -> 1
```
Nit: Move `based` out of the body of if and else to avoid having them duplicated twice in code.

dustanddreams replied:
Maybe at a later time. At the moment I prefer the two code paths to be clearly separate, and once there are no changes to make to the code generation, some factorization can be done.

kayceesrk commented on asmcomp/arm64/emit.mlp:
```diff
@@ -464,23 +464,46 @@ module BR = Branch_relaxation.Make (struct
-      and single = match memory_chunk with Single -> 2 | _ -> 1 in
-      based + barrier + single
+      if !Clflags.alternate_code_generation then begin
+        let based = match addressing_mode with Iindexed _ -> 0 | Ibased _ -> 1
```
Nit: Move `based` and `single` out of the body of if and else to avoid having them duplicated twice in code.

kayceesrk commented:
> My proposal: -barrier-placement=[fence-before-store | branch-after-load].

I agree with being explicit. However, the option applies only to Arm64 backend at the moment. Are there such architecture-specific configure time options already? Should configure reject this option on other architectures, show a warning or silently ignore them?

ghost commented:
> > My proposal: -barrier-placement=[fence-before-store | branch-after-load].
> 
> I agree with being explicit. However, the option applies only to Arm64 backend at the moment. Are there such architecture-specific configure time options already? Should configure reject this option on other architectures, show a warning or silently ignore them?

There is (or at least used to) a way to have backend-specific options, which won't be recognized anywhere else; I'll move the handling of that option to arm64 land.

kayceesrk commented:
@dustanddreams FYI, there is a new commit in https://github.com/ocaml/ocaml/pull/13393.

stedolan commented:
From a conversation I had with @kayceesrk at ICFP:

  - This looks like a good implementation of the BAL compilation scheme for the OCaml memory model (at least, once the extra commit on #13393 is picked up)
  - It's not obvious to me whether this is better or worse overall than the current one. This approach puts a fence on C calls rather than mutations. Which of these is more frequent varies by program.
  - I think more investigation into #13262 is needed. From some benchmarking at ICFP, it appears that M2 chips are sequentialising memory operations after `dmb ishld` under some poorly-understood conditions. (The slow variant of the benchmark slowed down from L1 access times per iteration to L2 access times per iteration when the data size outgrew L1, despite no obvious dependency between iterations except the fence. The fast variant of the benchmark also contained the fence but went faster than L2 access per iteration even on large data).
  
    If we can understand the conditions that trigger this behaviour (e.g. instruction alignment?) then we can possibly work around them (e.g. tactical NOPs?). If we don't understand the conditions that trigger this behaviour, then moving the fences around does not necessarily resolve the problem.

Commit:
Implement the arm64 alternate memory model instruction scheme.

This branch-after-load scheme will restrain from issueing memory barrier
instructions after stores, at the expense of an increase in code size.

Selection of the scheme is performed at compiler configuration time, using
the --enable-alternate-arm64-code-generation option.

Co-authored-by: KC Sivaramakrishnan <kc@kcsrk.info>

ghost commented:
Updated with KC's changes, moving the option to be arm64-only, and renaming it as suggested by Gabriel.

I have thought a bit on how to prevent mixing code objects produced with different states of the option.

One way would be to store this information in the object files, but this will be target format dependent (and thus might not be possible on macOS), and would require some assistance from the linker to reject conflicting files. Another way would be to use a different mangling scheme when BAL is used. This will however be a bit ugly as some symbols are referenced from `arm64.S` (e.g. `caml_system__code_begin) and thus must remain unchanged.

gasche commented:
To enforce compatibility between compilation options at runtime, I would put the information in the `.cmx` file (see file_formats/cmx_format.mli). I am not sure how to model architecture-specific linking options. Maybe just something like this?

```
type unit_infos = {
  ...
  mutable ui_configuration : configuration;
  ...
}

and configuration = {
  arm64_barrier_placement : arm64_barrier_placement option;
}
```

Besides: @stedolan had interesting comments on the need to explore the apple m1-specific performance cliffs we see, and the fact that branch-after-load may not protect ourselves from the issue (there is now another way to have a barrier-heavy loop, wich C calls, that may be similarly degraded). Is someone interested in actively looking into this?

ghost commented:
It would make sense to add information to the `cmx` files indeed. But wouldn't adding new fields require a change of magic numbers and regenerating the `boot` compilers as well?

gasche commented:
We only keep bytecode compilers in boot/, so there should be no change necessary there. But yes, .cmx magic files would need an update.

kayceesrk commented:
> Besides: @stedolan had interesting comments on the need to explore the apple m1-specific performance cliffs we see, and the fact that branch-after-load may not protect ourselves from the issue (there is now another way to have a barrier-heavy loop, wich C calls, that may be similarly degraded). Is someone interested in actively looking into this?

Stephen mentioned that we should get this bug report to ARM and/or Apple engineers for them to have a look. Perhaps @maranget has contacts at ARM who may know what to do with this information.

ghost commented:
Checkpoint of work in progress, `cmx` files now embed information and I'm working on adding code to prevent mixing incompatible cmx files at link time.

~~However my #.1 struggle at the moment is that I probably botched the magic number update and bootstrap, as trying to build from a clean tree fails, complaining about incorrect magic numbers for `boot/ocamlc`. It's very likely something I missed but I have not found that mistake yet.~~

ghost commented:
At this point, the strict rejection logic will prevent linking code compiled with `-mbarrier-placement=branch-after-load` against the stdlib, linked with the default `-mbarrier-placement=fence-before-store`.

I suppose this calls for an opam option to be able to pick the default value of this option at `configure` time. Are there any guidelines to follow when editing `ocaml-variants.opam`?

kayceesrk commented:
> At this point, the strict rejection logic will prevent linking code compiled with -mbarrier-placement=branch-after-load against the stdlib, linked with the default -mbarrier-placement=fence-before-store.
>
> I suppose this calls for an opam option to be able to pick the default value of this option at configure time. Are there any guidelines to follow when editing ocaml-variants.opam?

It is unsafe to mix `branch-after-load` and `fence-before-store`. And, we have a configure time option that changes the default `--enable-alternate-arm64-code-generation`: https://github.com/ocaml/ocaml/pull/13423/files#diff-49473dca262eeab3b4a43002adb08b4db31020d190caaad1594b47f1d5daa810R2470.

Isn't it simpler to only have the configure time option and completely remove the `ocamlopt` option. We can remove all of the cmx additions, relying on the guarantee that using a particular compiler variant only produces one kind of code. 

I expect that having the flexibility to compile code both ways using the same `ocamlopt` executable will rarely be needed. Managing different compiler switches is easy enough with opam. Hence, it wouldn't be a loss if this feature were removed.

ghost commented:
If you're positive that the opam option is good enough to prevent accidental mixing of both code flavours unless people are working really hard to shoot themselves in the feet, I'm all for removing the cmx format change and the `ocamlopt` compiler option.

Waiting a bit to let other people voice their opinion on that matter.

lthls commented:
I tend to agree with @kayceesrk. The only use I see for the `ocamlopt` option is to easily compare the output of both schemes on a given input, and I think that's only useful for compiler developers. Given that it's likely trivial to patch the compiler sources to temporarily restore this feature (just replace the constant in `asmcomp/arm64/arch.ml` by a `Sys.getenv` dispatch), I would be in favour of not exposing the feature to our users at all.

kayceesrk commented:
> If you're positive that the opam option is good enough to prevent accidental mixing of both code flavours unless people are working really hard to shoot themselves in the feet, I'm all for removing the cmx format change and the ocamlopt compiler option.

I think we have (or have had) a couple of cases where such accidental mixing may be dangerous, but they're only separated by configure time options -- frame pointers mode, flat float arrays, unsafe strings.

> Waiting a bit to let other people voice their opinion on that matter.

Indeed. Please do wait for others to chime in before undoing all the work.

dra27 commented:
The .cmx format change is an improvement, so I wouldn't remove that (i.e. it would be better after this PR to be switching frame pointers and flat arrays to do similar if necessary).

I agree with not exposing the option directly in ocamlopt at the moment, though - long-term, there should be one compiler to rule them all, but we're not there yet, and having an ocamlopt flag makes that far too easy for it to be appear in the flags fields of `dune` files, and foot shooting becomes very easy.

 So, yes, I think the scheme should be selected in `configure` (which in opam would be exposed by adding another `ocaml-option-` package) and then that compiler will only work with that scheme, but the rest of it should stay - in particular, that compiler should be able to complain if it's ever pointed at a .cmxa built by a compiler which is using the opposite scheme.

kayceesrk commented:
> So, yes, I think the scheme should be selected in configure (which in opam would be exposed by adding another ocaml-option- package) and then that compiler will only work with that scheme, but the rest of it should stay - in particular, that compiler should be able to complain if it's ever pointed at a .cmxa built by a compiler which is using the opposite scheme.

Sounds like a good balance to me.

Commit:
Teach opam about arm64 codegen option.

Commit:
Document this endeavour.

kayceesrk commented:
I can review the backend changes. Would someone else be willing to review the other changes?

ghost commented:
With that feedback, I've removed the `ocamlopt` option and kept everything else. The error message when attempting to mix conflicting objects might have become less understandable. Feel free to suggest something better.

lthls commented:
I find the cmx changes very unpleasing. I would expect at least a proper record with proper, non-string contents. But even with that, I just don't see the point.
If this is about having the configuration options in the cmx, then  just dump the `Config` module in some way, and that's it.
I'm actually surprised that the new option is not in `Config`, by the way. This means that you can't know which scheme your compiler uses just by doing `ocamlopt -config`, and it also means one additional `.in` file, for no good reason that I can see.

ghost commented:
The changes to the `.cmx` files have intentionally been as extensible as possible to let future changes not require changes to that format again.

Having the state of the option in `Config` is a good point. I am however quite reluctant to do changes there for something which is specific to a single platform.

nojb commented:
> The changes to the `.cmx` files have intentionally been as extensible as possible to let future changes not require changes to that format again.

I haven't been following the PR closely, but compilation artifacts are incompatible between different versions of the compiler (different magic number), so am not sure that "not requiring changes in the future" buys us anything here. Does it?

kayceesrk commented:
Stepping back a bit, there are two things in this PR

1. An alternative codegen for Arm64 backend chosen through a configure time parameter
2. cmx changes that prevent accidental mixing of artifacts compiled with incompatible options

(1) seems uncontentious and having only a configure time parameter is no worse than what we have with other configurations like support for frame pointers, safe string, etc. (2) seems contentious but only orthogonally related to (1). Should we split (1) and (2), to limit this PR to (1) and (2) can be discussed in a separate issue/draft PR where the purpose and the design can be debated?

avsm commented:
The configure-time parameter seems safe here, but bear in mind that OS packagers are unlikely to do the right thing by default, so it would be great to ensure that the default configure logic does the right thing. If we could (e.g.) detect an Apple Silicon chip in the autoconf logic and switch on the option by default, and have an option to pick a specific one as well, that should cover all the bases.  The default does the right thing, and a particular build can fix one of the compilation options (e.g. for the CI or a benchmarking rig).

The only case I'm wondering about where we may need the flexibility to dynamically select the strategy is cross-compilation.  Will a configure-time option scupper our ability to compile Linux/arm64 binaries from a macOS/arm64 host compiler or vice versa? This seems to fall under the (2) option of KC's rather than (1), so probably not a blocker at this stage, but something to bear in mind.

dra27 commented:
> I haven't been following the PR closely, but compilation artifacts are incompatible between different versions of the compiler (different magic number), so am not sure that "not requiring changes in the future" buys us anything here. Does it?

Indeed, but the point here is that these are incompatible artefacts between _the same version_ of the compiler. Magic numbers protect the marshalling format, and shouldn't be being used elsewhere (in particular, see @Octachron's and my comments in https://github.com/ocaml/ocaml/issues/13262#issuecomment-2301465861).

dra27 commented:
I agree that if the implementation in the .cmx file is contentious (I hadn't actually reviewed that part - it was the concept I was commenting on above) then it doesn't need to block the PR.

@avsm's point about auto-detection is well made, although is that possibly something to do in a subsequent version (I don't think we're yet _certain_ that the scheme should be being used on Apple silicon, right?). The cross-compilation bit is on the radar, yes, but this option I don't think is different from flambda/flat-float-array, etc. targets at the moment - the compilation scheme for the target is just as cross-compily as the architecture and OS (just that typically the target is also runnable on the host)

ghost commented:
Removed the `cmx` changes for now.

kayceesrk commented on configure.ac:
```diff
@@ -2462,6 +2464,15 @@ AS_IF([test x"$enable_frame_pointers" = "xyes"],
+arm64_barrier_placement_default=FenceBeforeStore
+AC_ARG_ENABLE([alternate-arm64-code-generation],
+  [AS_HELP_STRING([--enable-alternate-arm64-code-generation],
+    [default to -mbarrier-placement=branch-after-load on arm64])],
```
This option no longer exists.

```suggestion
    [default to BranchAfterLoad code generation on arm64])],
```

dustanddreams replied:
Fixed. Thanks for noticing!

kayceesrk commented on configure.ac:
```diff
@@ -2462,6 +2464,15 @@ AS_IF([test x"$enable_frame_pointers" = "xyes"],
+## arm64-specific options
+arm64_barrier_placement_default=FenceBeforeStore
+AC_ARG_ENABLE([alternate-arm64-code-generation],
+  [AS_HELP_STRING([--enable-alternate-arm64-code-generation],
```
Nit: Better to be explicit? 

```suggestion
  [AS_HELP_STRING([--enable-fence-before-store-arm64-codegen],
```

dustanddreams replied:
Changed to `--enable-branch-after-load-arm64-codegen` since `FenceBeforeStore` is the default.

kayceesrk reviewed (approved):
The current code LGTM. 

This PR will be discussed at the next dev meeting (28/10). Let's wait until then to decide about merging this change.

kayceesrk commented:
We discussed this at the developer meeting. The outcome of the discussion is that, there was no strong opposition, but we need a bit more work before we agree to merge this. Here are the reasons that were presented as reasons for this work not yet being ready:

1. Fence-before-store being slow on Apple Silicon may indicate a hardware bug. 
2. The design penalises programs that are "more functional" by moving the (hardware) fence from stores to being a (compiler) fence to loads. In typical programs, we expect loads to outnumber loads (dynamically). There is also a non-trivial code size impact.
3. The design adds fences to external calls, and it is not entirely clear that this is cheap. 
4. Unclear whether other Arm hardware exhibits similar issue.
5. There are also other schemes (data-dependency on loads), which should also be evaluated before we agree to merge this.

Next steps

1. @stedolan will attempt to reduce this to a small sequence of assembly instructions that exhibit this issue. We will then share this bug report to ARM/Apple.
2. More extensive evaluation on a variety of hardware (which @maranget may have at hand).
3. Evaluate alternative compilation scheme (data dependency on loads)

xavierleroy commented:
For a repro case that does not depend on OCaml (just C + assembly) : https://github.com/xavierleroy/arm64-barriers
